internal/venice-client-common/src/main/java/com/linkedin/venice/reliability/LoadController.java

```diff
@@ -0,0 +1,160 @@
+  private final int windowSizeInSec;
+  private final int rejectionRatioUpdateIntervalInSec;
+  private final double maxRejectionRatio;
+  private final Time time;
+  private final double acceptMultiplier;
+
+  private final Sensor requestSensor;
+  private final Metric requestMetric;
+  private final Sensor acceptSensor;
+  private final Metric acceptMetric;
+
+  private volatile long nextRejectionRatioUpdateTime = -1;
+  private volatile double rejectionRatio = 0;
+
+  private LoadController(Builder builder) {
+    this.windowSizeInSec = builder.windowSizeInSec;
+    this.rejectionRatioUpdateIntervalInSec = builder.rejectionRatioUpdateIntervalInSec;
+    if (this.rejectionRatioUpdateIntervalInSec >= this.windowSizeInSec) {
+      throw new IllegalArgumentException("Rejection ratio update interval should be less than window size");
+    }
+    this.maxRejectionRatio = builder.maxRejectionRatio;
+    this.acceptMultiplier = builder.acceptMultiplier;
+    this.time = builder.time;
+    MetricsRepository metricsRepository =
+        new MetricsRepository(new MetricConfig().timeWindow(windowSizeInSec, TimeUnit.SECONDS));
+    this.requestSensor = metricsRepository.sensor("request");
+    this.requestMetric = requestSensor.add("request", new SampledCount());
+    this.acceptSensor = metricsRepository.sensor("accept");
+    this.acceptMetric = acceptSensor.add("accept", new SampledCount());
+  }
+
+  public void recordRequest() {
+    requestSensor.record();
+  }
+
+  public void recordAccept() {
+    acceptSensor.record();
+  }
+
+  public double getRejectionRatio() {
+    if (time.getMilliseconds() < nextRejectionRatioUpdateTime) {
+      return rejectionRatio;
+    }
+    synchronized (this) {
+      if (time.getMilliseconds() < nextRejectionRatioUpdateTime) {
+        return rejectionRatio;
+      }
+      double requestCount = getMetricValue(requestMetric);
+      double acceptCount = getMetricValue(acceptMetric);
+      if (requestCount == 0.0d) {
+        rejectionRatio = 0;
+      } else {
+        // TODO: maybe we can remove `+1` since we have already checked whether `requestCount` is 0 or not before.
+        rejectionRatio = Math.max(0, (requestCount - acceptMultiplier * acceptCount) / (requestCount + 1));
+      }
+      nextRejectionRatioUpdateTime = time.getMilliseconds() + rejectionRatioUpdateIntervalInSec * 1000l;
+      rejectionRatio = Math.min(rejectionRatio, maxRejectionRatio);
+
+      return rejectionRatio;
+    }
+  }
+
+  public boolean isOverloaded() {
+    return getRejectionRatio() > 0;
+  }
+
+  private double getMetricValue(Metric metric) {
+    double value = metric.value();
+    return Double.isFinite(value) ? value : 0.0;
+  }
+
+  public boolean shouldRejectRequest() {
+    double rejectRatio = getRejectionRatio();
+    if (rejectRatio == 0) {
+      return false;
+    }
+    if (rejectRatio >= 1) {
+      return true;
+    }
+    return Math.random() < rejectRatio;
```

Comment:
As we employed 3 types of rejection in Store/Instance controller in client, rejection in Server host, is it possible that we will reject more agressively than we expected? If we assume each load controller will reject 50% of request, we will reject 1 - 50%^3 = 87.5% of requests.

Reply:
They have different focuses:
1. Server Controller is used to reject traffic when overloaded and these overloaded requests are used to send signal to the clients.
2. Fast Client will rely on these signals to decide how to reject the future traffic.
One important thing is that Server doesn't reject all the traffic when overloaded so Client won't shed all the requests.

And there is no 3 layers, but just two layers.

Quota rejection is a different category, which is for store overload, which can be different from server overload.

Regarding the rejection behaviors, we need to run multiple experiments with different params to evaluate them.